### PR TITLE
Storage/feature/stg87-trailing dot

### PR DESCRIPTION
**Pull request by microzchang**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp [hunk 1]
```diff
@@ -834,4 +834,157 @@ namespace Azure { namespace Storage { namespace Test {
...
+      auto shareClient = shareServiceClient.GetShareClient(shareName);
+      auto rootDirectoryClient = shareClient.GetRootDirectoryClient();
+      auto directoryClient
+          = rootDirectoryClient.GetSubdirectoryClient(directoryNameWithTrailingDot);
+
+      // Create
+      auto createResult = directoryClient.Create().Value;
+
+      // ListFilesAndDirectories
+      auto directories = rootDirectoryClient.ListFilesAndDirectories().Directories;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp:860]: This only lists the first page of results. The directory we want to check may not in this page, right?

File: sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp [hunk 2]
```diff
@@ -834,4 +834,157 @@ namespace Azure { namespace Storage { namespace Test {
...
+
+  TEST_F(FileShareDirectoryClientTest, AllowTrailingDot)
+  {
+    const std::string directoryName = RandomString();
+    const std::string directoryNameWithTrailingDot = directoryName + ".";
+    const std::string connectionString = StandardStorageConnectionString();
+    const std::string shareName = m_shareName;
+    auto options = InitStorageClientOptions<Files::Shares::ShareClientOptions>();
+
+    auto testTrailingDot = [&](bool allowTrailingDot) {
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp:846]: Besides explicit value (`true` and `false`), we also want to test the default behavior (not set).
This is very important to make sure there's not breaking change.

File: sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp [hunk 3]
```diff
@@ -834,4 +834,157 @@ namespace Azure { namespace Storage { namespace Test {
...
+      // Delete
+      EXPECT_NO_THROW(directoryClient.Delete());
+    };
+
+    // allowSourceTrailingDot = true
+    testTrailingDot(true);
+    // allowSourceTrailingDot = false
+    testTrailingDot(false);
+  }
+}}} // namespace Azure::Storage::Test
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp:990]: new line at the end of file

File: sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp (hunk 2 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp:846]: I think existing unit tests such as `FileShareFileClientTest::CreateDeleteFiles`, `FileShareFileClientTest::DownloadEmptyFile` will not set value when running the latest code. Is it necessary to add  a unit test to test this case(not set value)?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp:846]: File name in those test cases doesn't end with dot (`.`), does it?

Comment (microzchang) [reply sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp:846]: Yeah, I will add this test case.

File: sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp (hunk 1 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-shares/test/ut/share_directory_client_test.cpp]: In this case, there are only 2 elements in the list， but considering the integrity of the code, I think your comment is reasonable. I have updated the code to traverse the entire list.

**Review by Jinming-Hu — APPROVED** [review]

**Pull request merged**